File: backend/internal/models/survey.go
package models

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type TriggerType string

const (
	TriggerTypeManual    TriggerType = "manual"
	TriggerTypeLocation  TriggerType = "location"
	TriggerTypeTime      TriggerType = "time"
	TriggerTypeTripEnd   TriggerType = "trip_end"
	TriggerTypeStopStart TriggerType = "stop_start"
)

type QuestionType string

const (
	QuestionTypeYesNo          QuestionType = "yes_no"
	QuestionTypeMultipleChoice QuestionType = "multiple_choice"
	QuestionTypeNumber         QuestionType = "number"
	QuestionTypeText           QuestionType = "text"
)

type Survey struct {
	ID            uuid.UUID       `json:"id" db:"id"`
	Title         string          `json:"title" db:"title"`
	Description   *string         `json:"description,omitempty" db:"description"`
	TriggerType   TriggerType     `json:"trigger_type" db:"trigger_type"`
	TriggerConfig json.RawMessage `json:"trigger_config" db:"trigger_config"`
	IsActive      bool            `json:"is_active" db:"is_active"`
	CreatedAt     time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time       `json:"updated_at" db:"updated_at"`
}

type SurveyWithQuestions struct {
	Survey
	Questions []SurveyQuestion `json:"questions"`
}

type SurveyQuestion struct {
	ID           uuid.UUID       `json:"id" db:"id"`
	SurveyID     uuid.UUID       `json:"survey_id" db:"survey_id"`
	QuestionText string          `json:"question_text" db:"question_text"`
	QuestionType QuestionType    `json:"question_type" db:"question_type"`
	Options      json.RawMessage `json:"options,omitempty" db:"options"`
	IsRequired   bool            `json:"is_required" db:"is_required"`
	Order        int             `json:"order" db:"order_num"`
	CreatedAt    time.Time       `json:"created_at" db:"created_at"`
}

type SurveyResponse struct {
	ID         uuid.UUID  `json:"id" db:"id"`
	DriverID   uuid.UUID  `json:"driver_id" db:"driver_id"`
	SurveyID   uuid.UUID  `json:"survey_id" db:"survey_id"`
	QuestionID uuid.UUID  `json:"question_id" db:"question_id"`
	Answer     string     `json:"answer" db:"answer"`
	Latitude   *float64   `json:"latitude,omitempty" db:"latitude"`
	Longitude  *float64   `json:"longitude,omitempty" db:"longitude"`
	AnsweredAt time.Time  `json:"answered_at" db:"answered_at"`
}

type SurveyResponseWithDetails struct {
	SurveyResponse
	DriverName   string `json:"driver_name"`
	QuestionText string `json:"question_text"`
	SurveyTitle  string `json:"survey_title"`
}

// Request/Response DTOs
type SurveyCreateRequest struct {
	Title         string          `json:"title" binding:"required"`
	Description   *string         `json:"description,omitempty"`
	TriggerType   TriggerType     `json:"trigger_type" binding:"required"`
	TriggerConfig json.RawMessage `json:"trigger_config"`
	Questions     []QuestionCreateRequest `json:"questions" binding:"required,min=1"`
}

type QuestionCreateRequest struct {
	QuestionText string          `json:"question_text" binding:"required"`
	QuestionType QuestionType    `json:"question_type" binding:"required"`
	Options      json.RawMessage `json:"options,omitempty"`
	IsRequired   bool            `json:"is_required"`
	Order        int             `json:"order"`
}

type SurveySubmitRequest struct {
	Responses []ResponseSubmit `json:"responses" binding:"required"`
	Latitude  *float64         `json:"latitude,omitempty"`
	Longitude *float64         `json:"longitude,omitempty"`
}

type ResponseSubmit struct {
	QuestionID uuid.UUID `json:"question_id" binding:"required"`
	Answer     string    `json:"answer" binding:"required"`
}

// Trigger config yapıları
type LocationTriggerConfig struct {
	Province string `json:"province,omitempty"`
	District string `json:"district,omitempty"`
	Radius   int    `json:"radius,omitempty"` // metre cinsinden
}

type TimeTriggerConfig struct {
	Hour     int      `json:"hour"`
	Minute   int      `json:"minute"`
	Days     []string `json:"days"` // ["monday", "tuesday", ...]
}

type TripEndTriggerConfig struct {
	MinDistanceKm int `json:"min_distance_km,omitempty"`
	MinDurationMin int `json:"min_duration_min,omitempty"`
}

type StopStartTriggerConfig struct {
	LocationTypes []LocationType `json:"location_types,omitempty"`
	MinDurationMin int           `json:"min_duration_min,omitempty"`
}

File: mobile/lib/models/survey.dart
class Survey {
  final String id;
  final String title;
  final String? description;
  final String triggerType;
  final bool isActive;
  final List<SurveyQuestion> questions;
  final DateTime createdAt;

  Survey({
    required this.id,
    required this.title,
    this.description,
    required this.triggerType,
    required this.isActive,
    required this.questions,
    required this.createdAt,
  });

  factory Survey.fromJson(Map<String, dynamic> json) {
    return Survey(
      id: json['id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      triggerType: json['trigger_type'] as String? ?? 'manual',
      isActive: json['is_active'] as bool? ?? true,
      questions: (json['questions'] as List<dynamic>?)
              ?.map((q) => SurveyQuestion.fromJson(q as Map<String, dynamic>))
              .toList() ??
          [],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }
}

class SurveyQuestion {
  final String id;
  final String surveyId;
  final String questionText;
  final String questionType; // yes_no, multiple_choice, number, text
  final List<String>? options;
  final bool isRequired;
  final int order;

  SurveyQuestion({
    required this.id,
    required this.surveyId,
    required this.questionText,
    required this.questionType,
    this.options,
    required this.isRequired,
    required this.order,
  });

  factory SurveyQuestion.fromJson(Map<String, dynamic> json) {
    List<String>? options;
    if (json['options'] != null) {
      if (json['options'] is List) {
        options = (json['options'] as List).cast<String>();
      } else if (json['options'] is Map) {
        final optMap = json['options'] as Map<String, dynamic>;
        if (optMap['choices'] != null) {
          options = (optMap['choices'] as List).cast<String>();
        }
      }
    }

    return SurveyQuestion(
      id: json['id'] as String,
      surveyId: json['survey_id'] as String? ?? '',
      questionText: json['question_text'] as String,
      questionType: json['question_type'] as String? ?? 'text',
      options: options,
      isRequired: json['is_required'] as bool? ?? false,
      order: json['order'] as int? ?? 0,
    );
  }
}

class SurveyResponse {
  final String questionId;
  final String answer;

  SurveyResponse({
    required this.questionId,
    required this.answer,
  });

  Map<String, dynamic> toJson() {
    return {
      'question_id': questionId,
      'answer': answer,
    };
  }
}

class SurveySubmitRequest {
  final List<SurveyResponse> responses;
  final double? latitude;
  final double? longitude;

  SurveySubmitRequest({
    required this.responses,
    this.latitude,
    this.longitude,
  });

  Map<String, dynamic> toJson() {
    return {
      'responses': responses.map((r) => r.toJson()).toList(),
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
    };
  }
}
